### ISSUE TITLE
AudioKitUI: Include of non-modular header inside framework module

### ISSUE THREAD

**Issue by rob-k** [AudioKit/AudioKit#1454]
I'm trying to include AudioKitUI in my Swift project targeting iOS 10.0 (same for 11.0). Whenever I add `import AudioKitUI` to a Swift file and try building the project, I encounter multiple errors of the following type:
```
***/Pods/AudioKit/iOS/AudioKit.framework/Headers/EZAudio.h:55:9: Include of non-modular header inside framework module 'AudioKit.EZAudio': '***/Pods/AudioKit/iOS/AudioKit.framework/Headers/EZAudioUtilities.h'
```
Problematic headers seem to be `EZAudioUtilities.h` and `TPCircularBuffer.h`.

I added the library to my poject via Cocoapods:
```
pod 'AudioKit', '~> 4.0'
```

I already tried setting *Allow Non-modular Includes in Framework Modules*  to Yes in both my target and the Pods project with no success.

AudioKit itself is be imported with `import AudioKit` and used without any issues so far.

I am submitting this as an issue rather than a Stackoverflow question because I believe this is a problem within the framework itself or its Podspec.

**Comment by megastep** [comment]
Can you try to remove the `~> 4.0` part from your pod? You're probably just using an old version.

**Comment by rob-k** [comment]
To my knowledge, the [*optimistic operator*](https://guides.cocoapods.org/using/the-podfile.html) will lead to using the latest compatible version according to semver. So in my case, both `~> 4.0` and no version will lead to:
```
Installing AudioKit (4.2.4)
```

That being said, I removed the version operator from the Podfile, but the issue remains.

**Comment by aure** [comment]
Huh, optimistic, but scared of 3's?  Why didn't it get 4.3?

**Comment by rob-k** [comment]
Yes, that was my bad. I had to update my Repos by using `pod repo update` first to get the latest version. The issue remains though.

**Comment by maksutovic** [comment]
hi @rob-k I am having *exactly* the same issue. Was building/running my experiments with AudioKit with no issues whatsoever, and as soon as I wanted to use some of the UI components, I'm having the issue with non-modular headers.

For me the issue is with `EZPlot.h` line 27: `#import "AudioKit/EZAudioUtilities.h"`

I attempted the usual (now old) fixes like 
```
post_install do |installer|  
    installer.pods_project.build_configuration_list.build_configurations.each do |configuration|  
        configuration.build_settings['CLANG_ALLOW_NON_MODULAR_INCLUDES_IN_FRAMEWORK_MODULES'] = 'YES'  
    end  
end
```

but to no avail. 

My only concern is I am working on a project which has both Swift and Objective-C code, I was unsure if AudioKit did not support such an enviorment.

Is there a possible hotfix to moving the header to the .m file or is this on our end? Thanks so much

**Comment by thomjordan** [comment]
I haven't encountered this kind of issue with AudioKit yet, but I have
before when wrapping an Objective-C written framework inside of a new Swift
framework, most recently over this past weekend. This solution I found
worked very well and resolved the issue immediately, hopefully it can do
the same for you.

I adapted the instructions I found on StackOverflow here (also included
below):

https://stackoverflow.com/questions/32149371/import-objective-c-framework-into-swift-framework-google-analytics-cocoapod


For those who have continued to ask if I had solved this problem, the
following is how I accomplished this task importing the KochavaTracker SDK
Cocoapod into a framework. That SDK is currently written in Objective-C.
This answer is based on the answer provided by nuKs, with these specific
steps.

1) Create a folder named KochavaTracker under your project folder, i.e.
MyFramework/MyFramework/KochavaTracker.

2) Within that folder create a file named module.modulemap and insert the
following content:

/*

 This is the private module which is used to make private ObjC headers
available to Swift code.

 Note how all header files need to be specified with paths relative to this
file.


 This file lives inside a folder, and that folder is the actual module. In
Xcode the SWIFT_INCLUDE_PATHS needs to include the parent directory to that
folder.

 */

module KochavaTracker {

    header
"../../Pods/KochavaTrackeriOS/KochavaTrackeriOS/Classes/KochavaTracker.h"

    export *

}

This effectively creates a module which is a wrapper for interface of the
SDK, which Swift can later import like it does other Swift modules. Note
that it relies on a relative path to the Pods folder.

3) As found in build settings, modify your SWIFT_INCLUDE_PATHS to include:

$(SRCROOT)/MyFramework

This makes it so that you will be able to import the module and it will
locate it in/as the KochavaTracker folder which you created under it.

4) Add to your Swift code, where appropriate:

import KochavaTracker

From there you should be able to reference the classes in the
KochavaTracker module.

**Comment by ryangrimm** [comment]
I may be wrong, but I suspect that all of us in this thread that are having problems are using Objective-C in our app. Given the errors that I'm seeing, that seems to be central to the problem and why it doesn't show up in the MicrophoneAnalysis project for example (plus, the MicrophoneAnalysis project doesn't use a Podfile, it includes the framework directly). If I can find a few minutes I might take the MicrophoneAnalysis code, switch it over to using a Podfile and if it still works that way I'll include some Objective-C and see if that teases out the problem.

**Comment by ryangrimm** [comment]
Okay, my assumption about a project needing to have some portions written in Objective-C was incorrect. I created a new Swift project, copied the MicrophoneAnalysis files into it, added a Podfile, tried to build and am having the exact same two errors show up that I'm having in my own project.

The quick sample app I whipped up to show the problem can be found here: 

https://github.com/ryangrimm/MicrophoneAnalysis

Hopefully this simple test case will help the team figure out a solution.

**Comment by aure** [comment]
I'm going to assign this to @megastep and myself as something to resolve on next release.  I really want to make the Audiokit UI repo dynamic anyway, which will affect this.

**Comment by ryangrimm** [comment]
@aure I have a sample project that reproduces the issue here: https://github.com/ryangrimm/MicrophoneAnalysis

I did a `pod update` to grab the 4.5.1 version, opened the workspace in Xcode 10 and the problem remains.

**Comment by ryangrimm** [comment]
My Podfile.lock reports 4.5.1. I did a `pod install` just for safe measure. Cleaned (again). Build still fails:

```
/Users/test/MicrophoneAnalysis/Pods/AudioKit/iOS/AudioKitUI.framework/Headers/EZPlot.h:27:9: 
    error: include of non-modular header inside framework module 'AudioKitUI.EZPlot': 
    '/Users/test/MicrophoneAnalysis/Pods/Headers/Public/AudioKit/AudioKit/EZAudioUtilities.h'
#import "AudioKit/EZAudioUtilities.h"
```

**Comment by ryangrimm** [comment]
@aure What version of Xcode are you running? Perhaps that could explain the differences.

**Comment by aure** [comment]
I am running the latest version of Xcode10 and Mojave.  Do you want to do a screen share with me today?

**Comment by tdubik** [comment]
I have the same issue after upgrade to XCode 10.1. I tried clean-up of project and pods but compilation still fails with this error. Do you have any ideas how can I solve it ?

**Comment by pimisi** [comment]
The issue is not with the AudioKitUI but with how you are referencing it. When you have a framework that does not include objective-c bridging header you would have to use a modulemap file to gain access to you imported objective-c framework.

I had this issue with a different library and resolved it in the following way:

I'll use SampleFramework as my framework for this example and **AudioKitUI** as my objective-c framework

### My project directory:
```
SampleFramework
  - AudioKitUI // _Folder_
    - module.modulemap //  _NOTE: The name has to be module.modulemap not anything else_
  - AudioKitUI.h //  _File: You can name this file anything you wish, and must reference it with the same name_
  - SamplePublicHeader.h  
  - SampleFramework.h // _Must be the same name as the framework_
  - SampleSwiftClass.swift
```

### The contents of the files:

**module.modulemap**
```
module AudioKitUI {
    header "../AudioKitUI.h" // _This is the header file we have at SampleFrameWork/AudioKitUI.h. Notice the relative reference "../"_
    
    export *
}
```

**AudioKitUI.h**
```
#import <Foundation/Foundation.h>
#import AudioKitUI
```

**SamplePublicHeader.h** // _Not really necessary if you didn't include it in the 'SampleFramework.h' file because it is not needed_
```
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface SamplePublicHeader : NSObject

- (void)publicObjCMethod;

@end

NS_ASSUME_NONNULL_END
```

**SampleFramework.h**
```
#import <UIKit/UIKit.h>

//! Project version number for SampleFramework.
FOUNDATION_EXPORT double SampleFrameworkVersionNumber;

//! Project version string for SampleFramework.
FOUNDATION_EXPORT const unsigned char SampleFrameworkVersionString[];

// In this header, you should import all the public headers of your framework using statements like #import <SampleFramework/PublicHeader.h>
#import <SampleFramework/SamplePublicHeader.h>
```

**SampleSwiftClass.swift**
```
import AudioKitUI // _This will import your module, which will in-turn import the external module_

// _You can import any other module/framework you desire by including it in the AudioKitUI.h file as an import_

class SampleSwiftClass {
}
```
